File: examples/python/sim/runner.py
"""Per-session scenario runner.

Drives a single DetectDeepfake bidi RPC from a Scenario:
- emits CreateSessionResponse with a generated session_id,
- schedules curve samples + explicit events along the scenario's timeline,
- applies network latency/jitter on each emission,
- aborts the stream with the configured gRPC status if grpc.terminate_at_ms is set,
- drains the client's audio chunks in the background (content ignored),
- emits FinalResult on stream close (with the last computed score and the
  scenario's chosen label).
"""

from __future__ import annotations

import asyncio
import os
import random
import sys
import uuid
from dataclasses import dataclass
from typing import Any

import grpc

from aurigin.deepfake_detection.v1 import deepfake_detection_pb2 as pb

from . import curves
from .loader import Scenario

_STATUS_CODE_BY_NAME = {code.name: code for code in grpc.StatusCode}

# Opt-in per-emission logging. Always-on logs (start / fault / end) are below
# in run_session — this flag only gates the (potentially noisy) one-line-per-
# AnalysisResult output.
_LOG_ANALYSES = os.environ.get("SIM_LOG_ANALYSES", "").lower() in ("1", "true", "yes")


def _log(session_id: str, message: str) -> None:
    """Prefix every server-side log line with the session id for grep-ability.

    Writes to sys.stderr (not stdout) and bypasses Python's print() so
    process-manager wrappers (uv, hatch entry-point shims, just) can't
    silently buffer the destination. The startup line in server.py works
    the same way for the same reason.
    """
    sys.stderr.write(f"[{session_id}] {message}\n")
    sys.stderr.flush()

# Sentinel marking the end of the scheduled timeline. Consumed by the
# main generator loop to know it can stop pulling from the queue.
_TIMELINE_DONE = object()

# Bytes per sample for the AudioBuffer wire formats the deepfake-service
# decoder accepts. Used by _drain_audio's defensive duration-from-bytes
# fallback when the client doesn't populate duration_ns.
_BYTES_PER_SAMPLE = {"S16LE": 2, "F32LE": 4}


@dataclass
class _SessionState:
    session_id: str
    started_at: float                  # asyncio loop time
    accumulated_audio_ms: int = 0
    last_score: float = 0.0
    last_label: str = "bonafide"
    analysis_count: int = 0

    def now_ms(self, loop: asyncio.AbstractEventLoop) -> int:
        return int((loop.time() - self.started_at) * 1000)


def _make_session_id() -> str:
    return f"sim-{uuid.uuid4().hex[:8]}"


def _make_analysis_result(
    state: _SessionState,
    t_ms: int,
    score: float,
    label: str,
    confidence: float,
    duration_ms: int,
) -> pb.DetectDeepfakeResponse:
    state.last_score = score
    state.last_label = label
    state.analysis_count += 1
    return pb.DetectDeepfakeResponse(
        analysis_result=pb.AnalysisResult(
            audio_offset_ms=t_ms,
            duration_ms=duration_ms,
            score=score,
            label=label,
            confidence=confidence,
        ),
    )


async def _apply_network(network, rng: random.Random) -> None:
    delay_ms = network.base_latency_ms
    if network.jitter_ms:
        delay_ms += rng.randint(-network.jitter_ms, network.jitter_ms)
    if delay_ms > 0:
        await asyncio.sleep(delay_ms / 1000)


async def _drain_audio(request_iterator, state: _SessionState) -> None:
    """Consume audio chunks so the client can finish writing. Content ignored
    by the simulator — the scenario drives output, not the audio bytes."""
    logged_format = False
    async for msg in request_iterator:
        if msg.HasField("audio"):
            buf = msg.audio
            if not logged_format:
                _log(
                    state.session_id,
                    f"audio | format={buf.format or '?'} | rate={buf.rate} | channels={buf.channels}",
                )
                logged_format = True
            if buf.duration_ns > 0:
                state.accumulated_audio_ms += int(buf.duration_ns // 1_000_000)
            elif buf.rate and buf.channels:
                bps = _BYTES_PER_SAMPLE.get(buf.format, 2)
                state.accumulated_audio_ms += int(
                    len(buf.buffer) / bps / buf.channels / buf.rate * 1000
                )


async def _emit_timeline(
    scenario: Scenario,
    state: _SessionState,
    rng: random.Random,
    loop: asyncio.AbstractEventLoop,
    out_queue: asyncio.Queue,
) -> None:
    """Walk the scheduled timeline in order, sleeping between emissions."""
    timeline = _build_timeline(scenario)
    for at_ms, item in timeline:
        sleep_for = (at_ms / 1000.0) - (loop.time() - state.started_at)
        if sleep_for > 0:
            await asyncio.sleep(sleep_for)
        await _apply_network(scenario.network, rng)
        response = _materialise(item, scenario, state, rng, at_ms)
        if response is not None:
            await out_queue.put(response)
    await out_queue.put(_TIMELINE_DONE)


def _build_timeline(scenario: Scenario) -> list[tuple[int, dict[str, Any]]]:
    """Combine emissions (from backend_simulation or curve) + explicit events.

    Two scheduling modes:
      - **backend_simulation** (declarative): emissions are computed from
        (stream.duration_ms, analysis_interval_ms, min_chunk_duration_ms,
        tail_strategy). Mirrors how deepfake-service actually windows a
        finite-length audio stream. Curve still consulted for score values.
      - **curve.emit_every_ms** (scripted): emissions on a fixed wallclock
        grid driven by the curve. Pre-existing behavior; unchanged.

    Events are always overlaid on top.
    """
    items: list[tuple[int, dict[str, Any]]] = []

    bs = scenario.backend_simulation
    if bs is not None:
        items.extend(_emissions_from_backend_simulation(scenario))
    else:
        curve = scenario.confidence_curve
        if curve is not None:
            every = curve.get("emit_every_ms", 1000)
            for t in range(every, scenario.stream.duration_ms + 1, every):
                items.append((t, {"kind": "curve_sample", "at_ms": t}))

    for ev in scenario.events:
        items.append((ev.at_ms, {"kind": "event", "event": ev}))

    items.sort(key=lambda x: x[0])
    return items


def _emissions_from_backend_simulation(
    scenario: Scenario,
) -> list[tuple[int, dict[str, Any]]]:
    """Mirror dfs's window-emission rules over a finite stream.

    Walks the [0, duration_ms] interval in `analysis_interval_ms` steps and
    emits one `computed_emission` per main window. Residual after the last
    main window is handled per `tail_strategy`:

      - drop: residual < min_chunk_duration_ms → silently skipped
              residual ≥ min_chunk_duration_ms → emitted as a short tail
              window with duration_ms = residual
      - extend: residual < min_chunk_duration_ms → folded into the prior
                window (last main emission shifts to t=duration_ms with
                duration_ms = analysis_interval_ms + residual)
                residual ≥ min_chunk_duration_ms → same as drop's else branch
    """
    bs = scenario.backend_simulation
    assert bs is not None
    duration_ms = scenario.stream.duration_ms
    interval = bs.analysis_interval_ms
    min_chunk = bs.min_chunk_duration_ms
    silent_set = set(bs.silent_windows)
    n_main = duration_ms // interval
    residual = duration_ms - n_main * interval
    fold_residual = (
        residual > 0
        and residual < min_chunk
        and bs.tail_strategy == "extend"
        and n_main > 0
    )

    items: list[tuple[int, dict[str, Any]]] = []
    for i in range(1, n_main + 1):
        at_ms = i * interval
        window_dur = interval
        # Last main window absorbs the residual when extending.
        if i == n_main and fold_residual:
            at_ms = duration_ms
            window_dur = interval + residual
        items.append((at_ms, {
            "kind": "computed_emission",
            "at_ms": at_ms,
            "duration_ms": window_dur,
            "is_silent": i in silent_set,
            "silence_confidence": bs.silence_confidence,
        }))

    # Standalone tail emission (only when not folded into the prior window
    # and large enough to clear the min_chunk floor).
    if residual > 0 and not fold_residual and residual >= min_chunk:
        tail_idx = n_main + 1
        items.append((duration_ms, {
            "kind": "computed_emission",
            "at_ms": duration_ms,
            "duration_ms": residual,
            "is_silent": tail_idx in silent_set,
            "silence_confidence": bs.silence_confidence,
        }))

    return items


def _materialise(
    item: dict[str, Any],
    scenario: Scenario,
    state: _SessionState,
    rng: random.Random,
    at_ms: int,
) -> pb.DetectDeepfakeResponse | None:
    # AnalysisResult.duration_ms resolution order, finest grained wins:
    #   per-event payload.duration_ms > curve.analysis_window_ms > stream.chunk_interval_ms
    # Lets one scenario emit windows of varying lengths (e.g. tail_extended_
    # full_coverage where the last window is longer than the rest).
    curve = scenario.confidence_curve or {}
    curve_window_ms = curve.get("analysis_window_ms", scenario.stream.chunk_interval_ms)

    if item["kind"] == "computed_emission":
        # backend_simulation-driven emission. Silent windows skip the curve
        # entirely and emit a sentinel; otherwise curve (if any) computes
        # the score the same way curve_sample does.
        if item["is_silent"]:
            return _make_analysis_result(
                state,
                t_ms=at_ms,
                score=0.0,
                label="silence",
                confidence=float(item["silence_confidence"]),
                duration_ms=int(item["duration_ms"]),
            )
        if curve:
            score = curves.evaluate(curve, rng, at_ms)
            label = curves.label_for(curve, score)
        else:
            score = 0.0
            label = "bonafide"
        confidence = round(rng.uniform(0.85, 0.99), 3)
        return _make_analysis_result(
            state,
            t_ms=at_ms,
            score=score,
            label=label,
            confidence=confidence,
            duration_ms=int(item["duration_ms"]),
        )

    if item["kind"] == "curve_sample":
        score = curves.evaluate(curve, rng, at_ms)
        label = curves.label_for(curve, score)
        confidence = round(rng.uniform(0.85, 0.99), 3)
        return _make_analysis_result(
            state,
            t_ms=at_ms,
            score=score,
            label=label,
            confidence=confidence,
            duration_ms=curve_window_ms,
        )

    ev = item["event"]
    if ev.type in ("STREAM_STARTED", "STREAM_ENDED"):
        # Lifecycle markers are implicit on the wire (session/final messages).
        return None
    if ev.type == "ERROR":
        # Event-level errors are surfaced as a sentinel AnalysisResult with
        # label='error' and score 0; clients should treat label='error' as
        # non-actionable. (Future: dedicated proto message.)
        return _make_analysis_result(
            state,
            t_ms=at_ms,
            score=0.0,
            label="error",
            confidence=0.0,
            duration_ms=ev.payload.get("duration_ms", curve_window_ms),
        )
    # CONFIDENCE_UPDATE or FAKE_DETECTED — both map to AnalysisResult.
    payload = ev.payload
    score = float(payload.get("fake_probability", state.last_score))
    label = payload.get("label", curves.label_for(curve, score))
    confidence = float(payload.get("confidence", round(rng.uniform(0.85, 0.99), 3)))
    return _make_analysis_result(
        state,
        t_ms=at_ms,
        score=score,
        label=label,
        confidence=confidence,
        duration_ms=int(payload.get("duration_ms", curve_window_ms)),
    )


async def run_session(scenario: Scenario, request_iterator, context):
    """Async generator yielding DetectDeepfakeResponse for one bidi RPC."""
    loop = asyncio.get_running_loop()
    rng = random.Random(scenario.random_seed) if scenario.random_seed is not None else random.Random()

    # Set initial metadata if configured.
    if scenario.grpc.initial_metadata:
        await context.send_initial_metadata(list(scenario.grpc.initial_metadata.items()))

    # 1. Require CreateSessionRequest first.
    first = await request_iterator.__anext__()
    if not first.HasField("create_session_request"):
        await context.abort(grpc.StatusCode.INVALID_ARGUMENT, "Expected CreateSessionRequest first")
        return

    state = _SessionState(session_id=_make_session_id(), started_at=loop.time())
    yield pb.DetectDeepfakeResponse(
        create_session_response=pb.CreateSessionResponse(session_id=state.session_id),
    )
    _log(
        state.session_id,
        f"start | scenario={scenario.id} | duration_target={scenario.stream.duration_ms}ms"
        + (f" | seed={scenario.random_seed}" if scenario.random_seed is not None else ""),
    )  # verb width 5 chars — kept in sync with 'end  ' and 'fault' below

    # 2. Background tasks: drain audio, emit timeline, optional fault.
    out_queue: asyncio.Queue = asyncio.Queue()
    audio_task = asyncio.create_task(_drain_audio(request_iterator, state))
    emit_task = asyncio.create_task(_emit_timeline(scenario, state, rng, loop, out_queue))

    fault_task: asyncio.Task | None = None
    if scenario.grpc.terminate_at_ms is not None:
        fault_task = asyncio.create_task(asyncio.sleep(scenario.grpc.terminate_at_ms / 1000))

    try:
        # 3. Pump queue → wire until the timeline ends, the client closes
        #    its write side, or a fault fires.
        timeline_finished = False
        client_finished = False
        while not (timeline_finished and client_finished):
            queue_get = asyncio.create_task(out_queue.get())
            waiters = [queue_get]
            if not client_finished:
                waiters.append(audio_task)
            if fault_task:
                waiters.append(fault_task)

            done, _pending = await asyncio.wait(waiters, return_when=asyncio.FIRST_COMPLETED)

            if fault_task and fault_task in done:
                queue_get.cancel()
                code_name = scenario.grpc.status_code or "INTERNAL"
                code = _STATUS_CODE_BY_NAME.get(code_name, grpc.StatusCode.INTERNAL)
                message = scenario.grpc.status_message or "simulated fault"
                _log(
                    state.session_id,
                    f"fault | at={state.now_ms(loop)}ms | code={code_name} | message={message!r} "
                    f"| analyses_so_far={state.analysis_count}",
                )  # verb width 5 chars — kept in sync with 'start' and 'end  '
                await context.abort(code, message)
                return

            if audio_task in done and not client_finished:
                # Client closed the write side. Stop emitting and finalise.
                client_finished = True
                queue_get.cancel()
                emit_task.cancel()
                break

            item = await queue_get
            if item is _TIMELINE_DONE:
                timeline_finished = True
                continue
            if _LOG_ANALYSES and item.HasField("analysis_result"):
                r = item.analysis_result
                _log(
                    state.session_id,
                    f"analysis @ {r.audio_offset_ms / 1000:6.2f}s | score={r.score:.3f} "
                    f"| label={r.label} | confidence={r.confidence:.2f}",
                )
            yield item

        # 4. If the timeline finished first, still wait for the client to close.
        if not client_finished:
            await audio_task

        if scenario.grpc.trailing_metadata:
            context.set_trailing_metadata(list(scenario.grpc.trailing_metadata.items()))

        total_audio_ms = max(state.accumulated_audio_ms, scenario.stream.duration_ms)
        overall_label = state.last_label if state.analysis_count > 0 else "unknown"
        _log(
            state.session_id,
            f"end   | total={total_audio_ms}ms | analyses={state.analysis_count} "
            f"| score={state.last_score:.3f} | label={overall_label}",
        )  # 'end  ' padded to verb width 5 — matches 'start' and 'fault'
        yield pb.DetectDeepfakeResponse(
            final_result=pb.FinalResult(
                total_audio_ms=total_audio_ms,
                overall_score=state.last_score,
                overall_label=overall_label,
                analysis_count=state.analysis_count,
            ),
        )
    finally:
        for t in (audio_task, emit_task, fault_task):
            if t is not None and not t.done():
                t.cancel()

File: examples/typescript/sim/runner.ts
// Per-session scenario runner.
//
// Drives a single DetectDeepfake bidi RPC from a Scenario:
// - emits CreateSessionResponse with a generated session_id,
// - schedules curve samples + explicit events along the scenario's timeline,
// - applies network latency/jitter on each emission,
// - aborts the stream with the configured gRPC status if grpc.terminate_at_ms is set,
// - drains the client's audio chunks (content ignored),
// - emits FinalResult on stream close (with the last computed score and the
//   scenario's chosen label).
//
// Mirrors examples/python/sim/runner.py. Native asyncio doesn't have a direct
// equivalent in @grpc/grpc-js (which is event/callback-based), so the queue
// pump is implemented with setTimeout for scheduled emissions plus a single
// async tail to wait for client end / fault.

import { randomBytes } from "node:crypto";
import { Metadata, status as GrpcStatus, type ServerDuplexStream } from "@grpc/grpc-js";

import type {
  DetectDeepfakeRequest,
  DetectDeepfakeResponse,
} from "@aurigin/protos/aurigin/deepfake_detection/v1/deepfake_detection";

import * as curves from "./curves.js";
import type { Scenario, ScenarioEvent } from "./loader.js";

type Call = ServerDuplexStream<DetectDeepfakeRequest, DetectDeepfakeResponse>;
type Rng = curves.Rng;

// Opt-in per-emission logging. Always-on logs (start / fault / end) are
// emitted below in runSession — this flag only gates the (potentially noisy)
// one-line-per-AnalysisResult output.
const LOG_ANALYSES = ["1", "true", "yes"].includes(
  (process.env.SIM_LOG_ANALYSES ?? "").toLowerCase(),
);

function log(sessionId: string, message: string): void {
  // Prefix every server-side log line with the session id for grep-ability.
  // Use console.error (stderr) for parity with the Python side — process
  // supervisors are less likely to buffer stderr than stdout.
  console.error(`[${sessionId}] ${message}`);
}

const STATUS_CODE_BY_NAME: Record<string, number> = {
  OK: GrpcStatus.OK,
  CANCELLED: GrpcStatus.CANCELLED,
  UNKNOWN: GrpcStatus.UNKNOWN,
  INVALID_ARGUMENT: GrpcStatus.INVALID_ARGUMENT,
  DEADLINE_EXCEEDED: GrpcStatus.DEADLINE_EXCEEDED,
  NOT_FOUND: GrpcStatus.NOT_FOUND,
  ALREADY_EXISTS: GrpcStatus.ALREADY_EXISTS,
  PERMISSION_DENIED: GrpcStatus.PERMISSION_DENIED,
  RESOURCE_EXHAUSTED: GrpcStatus.RESOURCE_EXHAUSTED,
  FAILED_PRECONDITION: GrpcStatus.FAILED_PRECONDITION,
  ABORTED: GrpcStatus.ABORTED,
  OUT_OF_RANGE: GrpcStatus.OUT_OF_RANGE,
  UNIMPLEMENTED: GrpcStatus.UNIMPLEMENTED,
  INTERNAL: GrpcStatus.INTERNAL,
  UNAVAILABLE: GrpcStatus.UNAVAILABLE,
  DATA_LOSS: GrpcStatus.DATA_LOSS,
  UNAUTHENTICATED: GrpcStatus.UNAUTHENTICATED,
};

interface SessionState {
  sessionId: string;
  startedAt: number; // ms since process start (performance.now)
  accumulatedAudioMs: number;
  lastScore: number;
  lastLabel: string;
  analysisCount: number;
}

interface TimelineItem {
  kind: "curve_sample" | "event" | "computed_emission";
  atMs: number;
  event?: ScenarioEvent;
  // Set for kind=computed_emission (backend_simulation-driven).
  durationMs?: number;
  isSilent?: boolean;
  silenceConfidence?: number;
  // offsetMs usually equals atMs (fire-wallclock-time == audio offset under
  // the existing simulator convention), but tail_strategy='recompute'
  // slides it back so the offset reports where the slid window starts
  // rather than where it ended. Falls back to atMs when absent.
  offsetMs?: number;
}

// Bytes per sample for the AudioBuffer wire formats the deepfake-service
// decoder accepts. Used by the audio drainer's fallback duration-from-bytes
// calc when the client doesn't populate duration_ns.
const BYTES_PER_SAMPLE: Record<string, number> = { S16LE: 2, F32LE: 4 };

function makeSessionId(): string {
  // "sim-" prefix identifies this session as simulator-generated, matching
  // deepfake-service's "pre-" prefix (= prediction). Cross-cutting log
  // searches can tell at a glance whether a session id came from the real
  // backend or the scenario-driven simulator. 16 random bytes → 32 hex
  // chars (mirrors uuid4 hex; we don't need RFC 4122 compliance here).
  return `sim-${randomBytes(16).toString("hex")}`;
}

function mulberry32(seed: number): Rng {
  let a = seed >>> 0;
  return function (): number {
    a = (a + 0x6d2b79f5) >>> 0;
    let t = a;
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function applyNetwork(net: Scenario["network"], rng: Rng): Promise<void> {
  let delayMs = net.baseLatencyMs;
  if (net.jitterMs) {
    // Match Python's randint(-jitter, jitter) — integer in [-j, +j].
    const j = net.jitterMs;
    delayMs += Math.floor(rng() * (2 * j + 1)) - j;
  }
  if (delayMs > 0) {
    await sleep(delayMs);
  }
}

function buildTimeline(scenario: Scenario): TimelineItem[] {
  const items: TimelineItem[] = [];

  if (scenario.backendSimulation != null) {
    items.push(...emissionsFromBackendSimulation(scenario));
  } else {
    const curve = scenario.confidenceCurve;
    if (curve != null) {
      const every = (curve.emit_every_ms ?? 1000) as number;
      for (let t = every; t <= scenario.stream.durationMs; t += every) {
        items.push({ kind: "curve_sample", atMs: t });
      }
    }
  }

  for (const ev of scenario.events) {
    items.push({ kind: "event", atMs: ev.atMs, event: ev });
  }

  items.sort((a, b) => a.atMs - b.atMs);
  return items;
}

// Mirror dfs's window-emission rules over a finite stream. Walks
// [0, duration_ms] in `analysis_interval_ms` steps and emits one
// computed_emission per main window. Residual after the last main window
// follows `tail_strategy`:
//   - drop: residual < min_analysis_duration_ms → silently skipped
//           residual ≥ min_analysis_duration_ms → short tail window
//   - extend: residual < min_analysis_duration_ms → folded into prior window
//             (last emission shifts to t=duration_ms, longer duration)
//   - recompute: residual < min_analysis_duration_ms → last main emission
//                slides back to end at end-of-stream (offset shifts to
//                duration_ms - analysis_interval_ms, duration stays
//                analysis_interval_ms). Mirrors backend-app HTTP
//                /predict chunk_audio byte-for-byte.
//   - extend|recompute with residual ≥ min_analysis_duration_ms → same as
//     drop's else branch (standalone short tail).
function emissionsFromBackendSimulation(scenario: Scenario): TimelineItem[] {
  const bs = scenario.backendSimulation!;
  const durationMs = scenario.stream.durationMs;
  const interval = bs.analysisIntervalMs;
  const minAnalysis = bs.minAnalysisDurationMs;
  const silentSet = new Set(bs.silentWindows);
  const nMain = Math.floor(durationMs / interval);
  const residual = durationMs - nMain * interval;
  const canAbsorbTail =
    residual > 0 &&
    residual < minAnalysis &&
    nMain > 0 &&
    (bs.tailStrategy === "extend" || bs.tailStrategy === "recompute");

  const items: TimelineItem[] = [];
  for (let i = 1; i <= nMain; i++) {
    let atMs = i * interval;
    let windowDur = interval;
    let offsetMs = atMs;
    if (i === nMain && canAbsorbTail) {
      if (bs.tailStrategy === "extend") {
        atMs = durationMs;
        windowDur = interval + residual;
        offsetMs = atMs;
      } else {
        // recompute — fire at end-of-stream, but report the audio offset
        // as where the slid-back window starts (durationMs - interval).
        atMs = durationMs;
        windowDur = interval;
        offsetMs = durationMs - interval;
      }
    }
    items.push({
      kind: "computed_emission",
      atMs,
      durationMs: windowDur,
      isSilent: silentSet.has(i),
      silenceConfidence: bs.silenceConfidence,
      offsetMs,
    });
  }

  if (residual > 0 && !canAbsorbTail && residual >= minAnalysis) {
    const tailIdx = nMain + 1;
    items.push({
      kind: "computed_emission",
      atMs: durationMs,
      durationMs: residual,
      isSilent: silentSet.has(tailIdx),
      silenceConfidence: bs.silenceConfidence,
      offsetMs: durationMs,
    });
  }

  return items;
}

function makeAnalysisResult(
  state: SessionState,
  tMs: number,
  score: number,
  label: string,
  confidence: number,
  durationMs: number,
): DetectDeepfakeResponse {
  state.lastScore = score;
  state.lastLabel = label;
  state.analysisCount += 1;
  return {
    analysisResult: {
      audioOffsetMs: BigInt(tMs),
      durationMs: BigInt(durationMs),
      score,
      label,
      confidence,
    },
  } as DetectDeepfakeResponse;
}

function materialise(
  item: TimelineItem,
  scenario: Scenario,
  state: SessionState,
  rng: Rng,
  atMs: number,
): DetectDeepfakeResponse | null {
  // AnalysisResult.duration_ms resolution order, finest grained wins:
  //   per-event payload.duration_ms > curve.analysis_window_ms > stream.chunk_interval_ms
  // Lets one scenario emit windows of varying lengths (e.g. tail_extended_
  // full_coverage where the last window is longer than the rest).
  // Mirrors examples/python/sim/runner.py._materialise.
  const curve = scenario.confidenceCurve ?? {};
  const curveWindowMs = ((curve as any).analysis_window_ms as number | undefined) ?? scenario.stream.chunkIntervalMs;

  if (item.kind === "computed_emission") {
    // backend_simulation-driven emission. Silent windows skip the curve and
    // emit a sentinel; otherwise the curve (if any) computes the score the
    // same way curve_sample does.
    //
    // offsetMs is the wire audio_offset_ms. Usually equals atMs (fire-time ==
    // audio-offset under the existing simulator convention), but recompute
    // slides it back so the offset reports where the slid window starts.
    const emitOffset = item.offsetMs ?? atMs;
    const dur = item.durationMs ?? curveWindowMs;
    if (item.isSilent) {
      return makeAnalysisResult(
        state, emitOffset, 0, "silence", item.silenceConfidence ?? 0.95, dur,
      );
    }
    let score = 0;
    let label = "bonafide";
    if (scenario.confidenceCurve) {
      score = curves.evaluate(scenario.confidenceCurve, rng, atMs);
      label = curves.labelFor(scenario.confidenceCurve, score);
    }
    const confidence = Math.round((0.85 + rng() * 0.14) * 1000) / 1000;
    return makeAnalysisResult(state, emitOffset, score, label, confidence, dur);
  }

  if (item.kind === "curve_sample") {
    const c = scenario.confidenceCurve!;
    const score = curves.evaluate(c, rng, atMs);
    const label = curves.labelFor(c, score);
    const confidence = Math.round((0.85 + rng() * 0.14) * 1000) / 1000;
    return makeAnalysisResult(state, atMs, score, label, confidence, curveWindowMs);
  }

  const ev = item.event!;
  if (ev.type === "STREAM_STARTED" || ev.type === "STREAM_ENDED") {
    // Lifecycle markers are implicit on the wire (session/final messages).
    return null;
  }
  if (ev.type === "ERROR") {
    // Event-level errors are surfaced as a sentinel AnalysisResult with
    // label='error' and score 0; clients should treat label='error' as
    // non-actionable. (Future: dedicated proto message.)
    const dur = (ev.payload.duration_ms as number | undefined) ?? curveWindowMs;
    return makeAnalysisResult(state, atMs, 0, "error", 0, dur);
  }
  // CONFIDENCE_UPDATE or FAKE_DETECTED — both map to AnalysisResult.
  const payload = ev.payload;
  const score = (payload.fake_probability ?? state.lastScore) as number;
  const label = (payload.label ?? curves.labelFor(scenario.confidenceCurve ?? {}, score)) as string;
  const confidence = (payload.confidence ?? Math.round((0.85 + rng() * 0.14) * 1000) / 1000) as number;
  const dur = (payload.duration_ms as number | undefined) ?? curveWindowMs;
  return makeAnalysisResult(state, atMs, score, label, confidence, dur);
}

function audioChunkMs(audio: {
  durationNs?: bigint;
  rate?: number;
  channels?: number;
  buffer?: Uint8Array;
  format?: string;
}): number {
  const durationNs = audio.durationNs ?? 0n;
  if (durationNs > 0n) {
    return Number(durationNs / 1_000_000n);
  }
  const rate = audio.rate ?? 0;
  const channels = audio.channels ?? 0;
  const buf = audio.buffer;
  if (rate && channels && buf) {
    const bps = BYTES_PER_SAMPLE[audio.format ?? ""] ?? 2;
    return Math.round((buf.length / bps / channels / rate) * 1000);
  }
  return 0;
}

export async function runSession(scenario: Scenario, call: Call): Promise<void> {
  const rng = scenario.randomSeed != null ? mulberry32(scenario.randomSeed) : Math.random;

  // Initial metadata.
  if (Object.keys(scenario.grpc.initialMetadata).length > 0) {
    const md = new Metadata();
    for (const [k, v] of Object.entries(scenario.grpc.initialMetadata)) md.set(k, v);
    call.sendMetadata(md);
  }

  // 1. Require CreateSessionRequest first. Eagerly attach the data listener
  //    so we don't lose subsequent audio chunks while we await the first one.
  const state: SessionState = {
    sessionId: makeSessionId(),
    startedAt: performance.now(),
    accumulatedAudioMs: 0,
    lastScore: 0,
    lastLabel: "bonafide",
    analysisCount: 0,
  };

  let firstResolved = false;
  let firstResolve: (req: DetectDeepfakeRequest) => void;
  let firstReject: (err: unknown) => void;
  const firstMessage = new Promise<DetectDeepfakeRequest>((res, rej) => {
    firstResolve = res;
    firstReject = rej;
  });

  let clientEnded = false;
  let clientEndResolve: () => void;
  const clientEndedPromise = new Promise<void>((res) => {
    clientEndResolve = res;
  });

  let loggedFormat = false;
  call.on("data", (msg: DetectDeepfakeRequest) => {
    if (!firstResolved) {
      firstResolved = true;
      firstResolve(msg);
      return;
    }
    if (msg.audio) {
      if (!loggedFormat) {
        const a = msg.audio as any;
        log(
          state.sessionId,
          `audio | format=${a.format || "?"} | rate=${a.rate ?? 0} | channels=${a.channels ?? 0}`,
        );
        loggedFormat = true;
      }
      state.accumulatedAudioMs += audioChunkMs(msg.audio as any);
    }
  });

  call.on("end", () => {
    if (!firstResolved) {
      firstResolved = true;
      firstReject(new Error("Stream ended before CreateSessionRequest"));
    }
    clientEnded = true;
    clientEndResolve();
  });

  call.on("error", () => {
    // The transport surfaces errors via 'error'; treat as a client-end so we
    // tear down cleanly. The fault-injection path emits its own error after.
    clientEnded = true;
    clientEndResolve();
  });

  let first: DetectDeepfakeRequest;
  try {
    first = await firstMessage;
  } catch {
    return;
  }
  if (!first.createSessionRequest) {
    call.emit("error", {
      code: GrpcStatus.INVALID_ARGUMENT,
      message: "Expected CreateSessionRequest first",
    });
    return;
  }

  call.write({ createSessionResponse: { sessionId: state.sessionId } } as DetectDeepfakeResponse);
  // Verb width 5 chars — kept in sync with 'end  ' and 'fault' below so '|'
  // aligns when grepping per-session log output.
  log(
    state.sessionId,
    `start | scenario=${scenario.id} | duration_target=${scenario.stream.durationMs}ms` +
      (scenario.randomSeed != null ? ` | seed=${scenario.randomSeed}` : ""),
  );

  // 2. Schedule timeline emissions and fault. Each emission applies network
  //    delay before writing.
  const timeline = buildTimeline(scenario);
  let faultFired = false;
  const timers: NodeJS.Timeout[] = [];

  for (const item of timeline) {
    const t = setTimeout(async () => {
      if (clientEnded || faultFired) return;
      await applyNetwork(scenario.network, rng);
      if (clientEnded || faultFired) return;
      const response = materialise(item, scenario, state, rng, item.atMs);
      if (response) {
        if (LOG_ANALYSES && response.analysisResult) {
          const r = response.analysisResult;
          const offsetS = (Number(r.audioOffsetMs) / 1000).toFixed(2).padStart(6);
          log(
            state.sessionId,
            `analysis @ ${offsetS}s | score=${r.score.toFixed(3)} | ` +
              `label=${r.label} | confidence=${r.confidence.toFixed(2)}`,
          );
        }
        call.write(response);
      }
    }, item.atMs);
    timers.push(t);
  }

  if (scenario.grpc.terminateAtMs != null) {
    const t = setTimeout(() => {
      if (clientEnded) return;
      faultFired = true;
      const codeName = scenario.grpc.statusCode ?? "INTERNAL";
      const code = STATUS_CODE_BY_NAME[codeName] ?? GrpcStatus.INTERNAL;
      const message = scenario.grpc.statusMessage ?? "simulated fault";
      log(
        state.sessionId,
        // 'fault' is verb width 5 — kept in sync with 'start' / 'end  '.
        `fault | at=${scenario.grpc.terminateAtMs}ms | code=${codeName} | message=${JSON.stringify(message)} ` +
          `| analyses_so_far=${state.analysisCount}`,
      );
      call.emit("error", { code, message });
    }, scenario.grpc.terminateAtMs);
    timers.push(t);
  }

  // 3. Wait for the client to close, or for the timeline (+ stream duration)
  //    to fully elapse. Whichever comes first.
  const timelineEnd = sleep(scenario.stream.durationMs).then(() => "timeline" as const);
  const clientEnd = clientEndedPromise.then(() => "client" as const);
  await Promise.race([timelineEnd, clientEnd]);

  // If the timeline finished first, still wait for the client to close so we
  // drain remaining audio metadata and don't write FinalResult mid-stream.
  if (!clientEnded) {
    await clientEndedPromise;
  }

  // Stop any pending scheduled emissions.
  for (const t of timers) clearTimeout(t);

  if (faultFired) return;

  if (Object.keys(scenario.grpc.trailingMetadata).length > 0) {
    const md = new Metadata();
    for (const [k, v] of Object.entries(scenario.grpc.trailingMetadata)) md.set(k, v);
    // @grpc/grpc-js attaches trailers via the call's `sendMetadata`-like
    // mechanism through `call.end(metadata?)` — not all versions support it,
    // so we attach as best-effort to the response object instead.
    (call as any).trailingMetadata = md;
  }

  const totalAudioMs = Math.max(state.accumulatedAudioMs, scenario.stream.durationMs);
  const overallLabel = state.analysisCount > 0 ? state.lastLabel : "unknown";
  log(
    state.sessionId,
    // 'end  ' padded to verb width 5 — matches 'start' / 'fault'.
    `end   | total=${totalAudioMs}ms | analyses=${state.analysisCount} ` +
      `| score=${state.lastScore.toFixed(3)} | label=${overallLabel}`,
  );
  call.write({
    finalResult: {
      totalAudioMs: BigInt(totalAudioMs),
      overallScore: state.lastScore,
      overallLabel,
      analysisCount: state.analysisCount,
    },
  } as DetectDeepfakeResponse);
  call.end();
}
